### ISSUE TITLE
Feature request: Batched NUFFT calls

### ISSUE THREAD

**Issue by JSKenyon** [mreineck/ducc#40]
Hi @mreineck! 

I was wondering if you had any intention of supporting NUFFTs for batches of data/along an axis as is done in `finufft`? The specific use case is taking the NUFFT along a single axis of a data cube (with irregularly spaced coordinates). At present, I flatten all the non-transform dimensions and then use an explicit Python loop and a predefined plan object. I suspect that this is less efficient than it could be.

**Comment by mreineck** [comment]
I have been thinking about this for some time, but doing the batching properly at the innermost level of the routines is a really big change. To get good performance, the batches would have to be split into fixed-size "microbatches" of sizes 2, 4, 8 or similar, and this size depends on the dimensionality of the NUFFT, the data type, and the accuracy. I'm sure that it's possible to get imressive speedups for 1D NUFFTs, but for 2D the gains will probably be modest and for 3D batching might even be a bad idea.
Do you have some numbers how much the batching would pay off in your concrete situation using `finufft`? If that gives a considerable speedup I'd be happy to reconsider.

**Comment by JSKenyon** [comment]
Thanks for the quick response! I don't have real numbers to put to this, and I was only really thinking about the 1D case (I imagine wanting to take many 2D or 3D NUFFTs simultaneously is a less common use case). I was interested in knowing if the feature was planned because it would help me figure out why I don't see consistent CPU usage when running many single-threaded NUFFTs in parallel. This could stem from repeated acquisition of the GIL by the calling Python loops, or poor allocation handling in my code - I have only recently got things working for my specific application. 

For some context, I am using the NUFFT to convolve a 3D `(irregular_time, x, y)` image cube along the time axis. This is done in parallel by splitting the cube into many `(irregular_time, x', y')` chunks and running single-threaded `ducc` NUFFTs (nu2u and u2nu) on many chunks in parallel. It is this case which seems to struggle to exercise many CPU cores efficiently (but again, this may be due to many things e.g. disk I/O, cache thrashing, memory allocation). The batched transform would just allow me to get rid of the Python loops (and remove one potential source of slowdown).

**Comment by mreineck** [comment]
I think I see more clearly now! So the problem is not that the NUFFT in itself is slow, but the problem you are observing is more the overhead of calling it in parallel from Python? This could probably be fixed more easily than what I had in mind.

Can you give me rough numbers of nonuniform points and grid sizes for which you are seeing the inconsistencies?

**Comment by JSKenyon** [comment]
Let me try and throw together a Python reproducer for you. Doing so will let me check that I am not just being daft.

**Comment by mreineck** [comment]
Even better, thanks! :-)

**Comment by mreineck** [comment]
Maybe a related idea ...

I have been thinking about adding a method to the NUFFT plan that outputs the ideal ordering of the irregular points from the NUFFT's point of view. If the user provides the input data (resp. accepts the output data) in that order and states that by, say, setting `nu_sorted=True`, many NUFFTS can be done much more quickly,especially 1D ones, since this improves cache accesses a lot. Of course this puts the burden of re-ordering the nonuniform data on the caller...

**Comment by JSKenyon** [comment]
> Maybe a related idea ...
> 
> I have been thinking about adding a method to the NUFFT plan that outputs the ideal ordering of the irregular points from the NUFFT's point of view. If the user provides the input data (resp. accepts the output data) in that order and states that by, say, setting `nu_sorted=True`, many NUFFTS can be done much more quickly,especially 1D ones, since this improves cache accesses a lot. Of course this puts the burden of re-ordering the nonuniform data on the caller...

Hmm, I must admit that I am not intimately familiar with the NUFFT internals, so I am not sure what constitutes an ideal ordering. I think I would need to see it in practice to get a feel for how realistic it would be to reorder the cube appropriately.

Here is a reproducer - it looks quite long but I think it is relatively simple. This doesn't implement the convolution, it simply takes the data from non-uniform to uniform and back again. If you see any other glaring errors in the code, please let me know! 

```python
import numpy as np
from ducc0.nufft import plan
from concurrent.futures import ThreadPoolExecutor, as_completed


def roundtrip(signal, signal_coords, epsilon=2e-6, nthreads=1):

    delta = np.diff(signal_coords).min()
    domain = (signal_coords.max() - signal_coords.min()) + delta
    n_modes = int(domain / delta)

    nu2u_signal = np.empty((n_modes,), dtype=np.complex128)
    u2nu_signal = np.empty_like(signal, dtype=np.complex128)

    nu2u_plan = plan(
        nu2u=True,
        coord=signal_coords[:, None],
        grid_shape=(n_modes,),
        epsilon=epsilon,
        nthreads=nthreads,
        periodicity=domain
    )

    u2nu_plan = plan(
        nu2u=False,
        coord=signal_coords[:, None],
        grid_shape=(n_modes,),
        epsilon=epsilon,
        nthreads=nthreads,
        periodicity=domain
    )

    for i in range(signal.shape[0]):
        nu2u_plan.nu2u(
            points=signal[i],
            forward=True,
            out=nu2u_signal,
        )

        u2nu_plan.u2nu(
            grid=nu2u_signal,
            out=u2nu_signal[i],
            forward=False,
        )

    nu2u_unity = nu2u_plan.nu2u(
        points=np.ones_like(signal_coords).astype(np.complex128),
        forward=True,
        out=np.empty((n_modes,), dtype=np.complex128),
    )

    u2nu_norm = u2nu_plan.u2nu(
        grid=nu2u_unity,
        forward=False,
    )
    
    return u2nu_signal/u2nu_norm[None, :]


if __name__ == "__main__":
    np.random.seed(1)

    n_thread = 32

    image_dims = (n_t, n_x, n_y) = (2048, 2048, 2048)
    chunk_dims = (_, n_xp, n_yp) = (n_t, 128, 128)
    flat_chunk_dims = (n_xp * n_yp, n_t)
    n_chunk = (n_x // n_xp) * (n_y // n_yp) 

    times = np.arange(n_t, dtype=np.float64)
    times += np.random.uniform(-0.1, 0.1, size=times.shape)
    signal = np.ones(flat_chunk_dims, dtype=np.complex128) 
    signal += np.random.normal(scale=0.1, size=signal.shape)

    with ThreadPoolExecutor(max_workers=n_thread) as executor:
        futures = {
            executor.submit(roundtrip, signal, times): i for i in range(n_chunk)
        }

        for future in as_completed(futures):
            print(f"Finished computing roundtrip {futures[future]}.")
            futures.pop(future)
```

For the settings above, this typically produces htop output like:
![image](https://github.com/user-attachments/assets/275992c9-7e1f-4411-8458-d57df6a14bf8)

**Comment by mreineck** [comment]
Hmm, maybe it's the way how Python assigns tasks to threads ... you are asking for 32 threads, but Python's thread pool may have size 64 (matching the number of virtual CPUs on the machine). If it then assigns tasks round robin, things may become weird. But this is pure speculation.

**Comment by mreineck** [comment]
I think I understand the issue now ... the problem is (I think) that a large part of the thread-parallel block (i.e. `roundtrip`) is actually written in Python, and the Python part is executed serially, because of the GIL.Even if most of the computational work is done inside C++, after every single 1D NUFFT every thread has to reacquire the GIL and this causes a massive bottleneck.
Having a batched NUFFT in C++, even if it is implemented very naively, would help a lot here, I agree. On the other hand, Python is slowly getting rid of the GIL starting with version 3.13 and the situation will hopefully improve by itself fairly soon.
I will definitely keep batched NUFFTs on my feature list, but I cannot give any promises, since the list is already rather long and I have to work on different projects during the next months :-/

**Comment by JSKenyon** [comment]
> I think I understand the issue now ... the problem is (I think) that a large part of the thread-parallel block (i.e. `roundtrip`) is actually written in Python, and the Python part is executed serially, because of the GIL.Even if most of the computational work is done inside C++, after every single 1D NUFFT every thread has to reacquire the GIL and this causes a massive bottleneck. Having a batched NUFFT in C++, even if it is implemented very naively, would help a lot here, I agree. On the other hand, Python is slowly getting rid of the GIL starting with version 3.13 and the situation will hopefully improve by itself fairly soon. I will definitely keep batched NUFFTs on my feature list, but I cannot give any promises, since the list is already rather long and I have to work on different projects during the next months :-/

Yeah - GIL contention is consistent with my instincts. There is no rush from my side - I am happy to hear that it is on the list, regardless of whether it will be implemented in the near future. Thanks for the feedback!

**Comment by JSKenyon** [comment]
As a follow-up, I still see poor resource use when replacing the `ThreadPoolExecutor` with a `ProcessPoolExecutor`. This is somewhat mysterious to me as it should eliminate any GIL related problems.

**Comment by mreineck** [comment]
Doesn't `ProcessPoolExecutor` imply having several Python interpreters running and copying the relevant data between those, MPI-style? I'm not sure how much overhead that will add, but it should be noticeable.

**Comment by JSKenyon** [comment]
You are absolutely correct - by moving the array creation into the `roundtrip` function, the multiprocessing approach works as expected i.e. seems to adequately use all the cores. This likely means that that the serialization of the data between processes is the real bottle neck in the multiprocessing case. This might help me solve my original problem - thanks for all your help. I will leave the issue open as it would still be useful to have the batched NUFFTs to avoid the GIL problem when using Python multithreading.

**Comment by mreineck** [comment]
My hope is that even without any change to `ducc` your multithreaded version will behave much better on a free-threaded Python interpreter (https://peps.python.org/pep-0703/). Unfortunately, this can currently only be tested by custom compiling CPython 3.13, which I'm not brave enough to do...
In any case, I'm happy to leave the issue open!

**Comment by mreineck** [comment]
Small update on this: I have experimental support for batching on the `nufft_split` branch. It can be enabled by simply calling the transforms with a prepended axis on the point and grid arrays.
Please note that the internal parallelization is still happening per tansform, i.e. if you have a lot of very quick transforms, then using many threads will probably not gain you much. I'll try to improve this in the future.

**Comment by JSKenyon** [comment]
Thanks for the update @mreineck! I will try and get round to trying it out!

**Comment by mreineck** [comment]
This is now part of `ducc0` 0.36.

**Comment by landmanbester** [comment]
Quick question about this, does this obviate the need to pre-compute a plan for repeated transforms of the same type? I mean, the plan is computed once and reused for all transforms right?

**Comment by mreineck** [comment]
If you can afford (memory might be a bottleneck here) to do several transforms in a single call, then yes, a plan will be built internally and used for these transforms, i.e. the precomputations will be done just once.

**Comment by landmanbester** [comment]
I'm not sure if I am doing something silly here but I can't figure out why the following
```python
import numpy as np
from ducc0.nufft import nu2u

ntrans = 111
npoints = 222
nthreads = 1
nmodes = 1024
epsilon = 1e-4
nthreads=1
domain = 2.0
signal = np.random.rand(ntrans, npoints).astype(np.complex64)
signal_coords = np.random.rand(npoints)[:, None].astype(np.float32)
nufft_signal = np.zeros((ntrans, nmodes), dtype=np.complex64)

nu2u(points=signal,
    coord=signal_coords[:, None],
    forward=True,
    out=nufft_signal,
    epsilon=epsilon,
    nthreads=nthreads,
    periodicity=domain)
```

fails with

```bash
Traceback (most recent call last):
  File "/home/bester/software/scratch/nufft_ducc.py", line 15, in <module>
    nu2u(points=signal,
RuntimeError: 
./src/ducc0/bindings/pybind_utils.h: 98 (std::array<long int, ndim> ducc0::detail_pybind::copy_fixstrides(const pybind11::array&, bool) [with T = float; long unsigned int ndim = 2]):

Assertion failure
incorrect number of dimensions
```

What I am doing wrong here?

**Comment by mreineck** [comment]
I think it's because you are using  `[:, None]` twice to append a dimension to `signal_coords`. If I remove this in the actual call, things seem to work for me.

**Comment by landmanbester** [comment]
Ah, yes. Thanks

**Comment by mreineck** [comment]
I have a idea how to improve the error message here; but this requires a lot of manual work and will take time.

**Comment by mreineck** [comment]
In the future, the exception messages from within the C++ code should (in most cases) start with the name of the respective parameter; it's not perfect, but it should help narrow down the search.

**Comment by landmanbester** [comment]
Ah, that would already be very useful! Thanks @mreineck
